### PR TITLE
gh-118928: sqlite3: disallow sequences of params with named placeholders

### LINKED ISSUE #118928: sqlite3: disallow using a sequence of params with named placeholders
# Feature or enhancement

Follow-up of...:

- #101693:

> Based on the discussion in the above linked Discourse topic, I propose to now issue a deprecation warning if sequences are used with named placeholders. The deprecation warning should inform that from Python 3.14 and onward, sqlite3.ProgrammingError will be raised instead.

3.14 dev is now opened, so let's convert the deprecation warning to a programming error sooner than later.

### Linked PRs
* gh-118929
* gh-119197
* gh-119208
* gh-135163
* gh-135342

### PR DESCRIPTION
**Pull request by erlend-aasland**
Follow-up of gh-101693.

* Issue: gh-118928

----
📚 Documentation preview 📚: https://cpython-previews--118929.org.readthedocs.build/

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. @felixxm

**Comment by erlend-aasland** [comment]
@hugovk, would you mind taking a look at the What's New entry?

File: Doc/whatsnew/3.14.rst [hunk 1]
```diff
@@ -142,6 +142,11 @@ sqlite3
 * Remove :data:`!version` and :data:`!version_info` from :mod:`sqlite3`.
   (Contributed by Hugo van Kemenade in :gh:`118924`.)
 
+* Disallow using a sequence of params with named placeholders.
+  This had previously raised a :exc:`DeprecationWarning` since Python 3.12;
+  it will now raise a :exc:`sqlite3.ProgrammingError`.
+  (Contributed by Erlend E. Aasland in :gh:`118928` and :gh:`101693`.)
```
Comment (hugovk) [inline Doc/whatsnew/3.14.rst:148]: We don't usually include the original issue where it was first deprecated, but it's not a bad idea.

File: Doc/whatsnew/3.14.rst [hunk 2]
```diff
@@ -142,6 +142,11 @@ sqlite3
 * Remove :data:`!version` and :data:`!version_info` from :mod:`sqlite3`.
   (Contributed by Hugo van Kemenade in :gh:`118924`.)
 
+* Disallow using a sequence of params with named placeholders.
```
Comment (hugovk) [inline Doc/whatsnew/3.14.rst]:
```suggestion
* Disallow using a sequence of parameters with named placeholders.
```

File: Misc/NEWS.d/next/Library/2024-05-10-22-36-01.gh-issue-118928.IW7Ukv.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Disallow using a sequence of params with named placeholders in
```
Comment (hugovk) [inline Misc/NEWS.d/next/Library/2024-05-10-22-36-01.gh-issue-118928.IW7Ukv.rst]:
```suggestion
Disallow using a sequence of parameters with named placeholders in
```

**Review by hugovk — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks, Hugo!

**Review by felixxm — APPROVED** [review]
@erlend-aasland Thanks :+1:

**Comment by erlend-aasland** [comment]
Thanks, Mariusz!

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks, Serhiy. I made the diff slightly nicer with 642015ec711b2d534ddece6d5411612abaa7f354. Enabling auto-merge.

**Pull request merged**